lib/libisns/isns.3
.\"     $NetBSD: isns.3,v 1.2 2011/01/16 08:00:01 wiz Exp $
.\"
.\" Copyright (c) 2004,2009 The NetBSD Foundation, Inc.
.\" All rights reserved.
.\"
.\" This code is derived from software contributed to The NetBSD Foundation
.\" by Wasabi Systems, Inc.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
.\" ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
.\" TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
.\" BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.Dd October 1, 2009
.Dt ISNS 3
.Os
.Sh NAME
.Nm isns
.Nd iSNS protocol support library
.Sh LIBRARY
.Lb libisns
.Sh SYNOPSIS
.In isns.h
.Ft int
.Fn isns_init "ISNS_HANDLE *h" "int is_server"
.Ft void
.Fn isns_stop "ISNS_HANDLE h"
.Ft int
.Fn isns_add_servercon "ISNS_HANDLE h" "int fd" "struct addrinfo *ai"
.Ft int
.Fn isns_init_reg_refresh "ISNS_HANDLE h" "const char *node" "int interval"
.Ft ISNS_TRANS
.Fn isns_new_trans "ISNS_HANDLE h" "uint16_t func_id" "uint16_t pdu_flags"
.Ft int
.Fn isns_add_tlv "ISNS_TRANS t" "uint32_t tag" "int data_len" "const void *data_p"
.Ft int
.Fn isns_add_string "ISNS_TRANS t" "uint32_t tag" "const char *s"
.Ft int
.Fn isns_send_trans "ISNS_TRANS t" "const struct timespec *timeout_p" "uint32_t *status_p"
.Ft int
.Fn isns_get_tlv "ISNS_TRANS t" "int which_tlv" "uint32_t *tag_p" "int data_len_p" "void **data_pp"
.Ft void
.Fn isns_free_trans "ISNS_TRANS t"
.Sh DESCRIPTION
The
.Nm
library exports an API that simplifies Internet Storage Name
Service (iSNS) client implementations.
The API defines a transactional model in support of:
.Pp
.Bl -bullet -width 3n -offset indent -compact
.It
generating iSNS messages composed of iSNS attributes expressed in
Tag-Length-Value (TLV) data format
.It
submitting iSNS Protocol (iSNSP) messages
.It
optionally waiting for iSNSP responses
.El
.Pp
.Nm
does not currently support receipt of iSNS Heartbeat messages, State Change
Notification (SCN) messages, or Entity Status Inquiry (ESI) messages.
.Sh INITIALIZATION
An iSNS client that uses
.Nm
must call
.Fn isns_init
to initialize the iSNS environment.
This call will create a thread to handle client-server communication, and
as such should only be called when thread creation is appropriate (such
as after a daemonized program forks).
.Pp
The value passed as
.Ar is_server
is used to set iSNSP message format Flags
"Sender is the iSNS client" (bit position 16) and "Sender is the iSNS server"
(bit position 17).
For now the value 0 (zero) should be passed for
.Ar is_server .
The value returned in
.Ar h
should be considered opaque by the caller.
This value is passed unchanged to
.Fn isns_add_servercon ,
.Fn isns_init_reg_refresh ,
.Fn isns_stop ,
and
.Fn isns_new_trans .
.Pp
.Fn isns_stop
should be called when the iSNS environment is no longer needed.
This call will kill any threads created by
.Fn isns_init .
.Sh CONFIGURATION
Following initialization,
.Fn isns_add_servercon
should be used to make the iSNS environment aware of the iSNS
server to which iSNSP queries and requests are to be sent.
This routine should not be called by a program acting as an iSNS server.
.Pp
A connected TCP socket descriptor is passed as parameter
.Ar fd .
Parameter
.Ar ai
is the address of the remote TCP endpoint.
It is included so that reconnection may be attempted by
.Nm
in the event that the TCP connection is broken.
.Pp
Certain iSNS servers will limit registration lifetimes, and will
refresh registrations after any request from a given iSNS entity.
The
.Fn isns_init_reg_refresh
function offers a way for
.Nm
to refresh registrations on behalf of the iSNS client.
.Pp
Parameter
.Ar node
is the
.Dq iSCSI Name
attribute used for the periodic queries.
It should be the name of an iSCSI node within the registered iSNS entity.
The desired refresh interval, in seconds, is passed in parameter
.Ar interval .
.Sh TRANSACTIONS
.Fn isns_new_trans
creates new iSNS transactions.
.Pp
Parameter
.Ar func_id
is used as the iSNSP message id.
Parameter
.Ar pdu_flags
is used to set iSNSP message format Flags and is
exposed to allow callers to set flag "Replace flag" (bit position 19).
This provides callers with a way
to specify whether a Device Attribute Registration Request is intended to
update or replace an existing registration.
This is currently the only use defined for parameter
.Ar pdu_flags .
.Pp
Once a new transaction has been created, callers can specify iSNS attributes
used for registration and query requests.
TLV data may be added using either
.Fn isns_add_tlv
or
.Fn isns_add_string .
.Pp
Parameter
.Ar tag
is the iSNS Tag corresponding to the attribute being added.
Parameter
.Ar data_len
is the length of the attribute value.
Parameter
.Ar data_p
references the attribute value.
The caller does not need to handle iSNS attribute 4-byte alignment requirements.
This is handled by the iSNS environment on behalf of the caller.
.Fn isns_add_string
may be used if the attribute value is a NUL terminated C string.
.Pp
Once a transaction has been populated with any required TLV data,
.Fn isns_send_trans
can be used to submit an iSNSP registration or query message.
.Pp
Callers that submit iSNSP query messages may need to wait for returned data.
.Fn isns_send_trans
supports bounded waits.
Successful waits, those that do not time out, return the iSNSP response
status code received in the iSNSP response message.
If a wait does time out, the value of
.Ar status_p
is undefined.
Callers that do not need to wait for returned data can simply
pass
.Dv NULL
for parameter
.Ar timeout_p .
Callers should set parameter
.Ar status_p
to
.Dv NULL
if not waiting.
.Pp
.Fn isns_get_tlv
is used to retrieve TLV data returned in a transaction.
The first call to
.Fn isns_get_tlv
should pass the value
.Dv ISNS_TLV_FIRST
for parameter
.Ar which_tlv .
Each subsequent TLV can be retrieved by passing in
.Dv ISNS_TLV_NEXT
in place of
.Dv ISNS_TLV_FIRST .
.Pp
When a caller is done with a transaction, having submitted either a
registration or a query message and retrieved any returned TLV data,
.Fn isns_free_trans
should be used to release resources used by the transaction.
.Sh RETURN VALUES
.Fn isns_init ,
.Fn isns_add_servercon ,
.Fn isns_init_reg_refresh ,
.Fn isns_add_tlv ,
.Fn isns_add_string ,
and
.Fn isns_send_trans
return 0 on success, or \-1 on failure.
.Fn isns_new_trans
returns 0 on success, or
.Dv ISNS_INVALID_TRANS
on failure.
.Fn isns_get_tlv
returns 0 on success, or
.Er ENOENT
if there are no TLVs to retrieve.
.Sh HISTORY
.Nm
first appeared in
.Nx 6.0 .
The
.Nm
implementation was contributed to the
.Nx
Foundation by Wasabi Systems, Inc.


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.1494 $>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 5.0 to NetBSD 6.0:

	syslog(3): Import Martin Schuette's Summer of Code project:
		Reliable TCP connections, syslog protocol API,
		encryption. [christos 20081031]
	agp(4): Add support for Intel G35. [christos 20081108]
	kernel: Add generic producer/consumer queue. [matt 20081111]
	lkm: Removed. Superseded by the new module(7) framework. [ad 20081112]
	module(7): Allow the POSIX semaphore code to be loaded as a module.
		[ad 20081112]
	module(7): Make compat_ossaudio a module. [ad 20081113]
	module(7): Make compat_ndis a module. [ad 20081113]
	build.sh: Add support for new MAKEVERBOSE levels 3 and 4.
		[apb 20081113]
	libfetch: Update to libfetch 2.20. Allow HTTP auth to be provided in
		URL. [joerg 20081114]
	module(7): Make POSIX AIO loadable as a module. [ad 20081114]
	amd64, i386: Make boot loader autoload a kernel module corresponding
		to the root file system type. [ad 20081119]
	module(7): Make the emulations, exec formats, coredump, NFS, and
		the NFS server into modules. [ad 20081119]
	rump(3): Move VFS functionality away from librump to a new library:
		librumpvfs [pooka 20081119]
	kernel, libc: Add new ptree (Patricia / RADIX tree) implementation.
		[matt 20081120]
	i386: Switch the i386 installation CDs away from RAM disks and
		instead use the CD-ROM as the root file system. [ad 20081121]
	i386: Remove most compat and file system options from GENERIC
		since they are also provided by modules. [ad 20081122]
	nfs: Avoid noisy nfs_timer/nfs_reply DEBUG output that occurs
		when the NFS server goes away. [mrg 20081123]
	compat_osf1(8): Add getpgid, needed for Tru64 4.0g ksh.
		[njoly 20081124]
	module(7): Make PPP compressors modules. [cube 20081125]
	i386: Make pxeboot(8) understand boot.cfg(5). [ad 20081125]
	agp(4): Add support for the Intel G45 AGP. [christos 20081129]
	compat_osf1(8): Fix getdirentries(2). [njoly 20081202]
	module(7): Add a couple of sysctl(8) entries, kern.module.autoload and
		kern.module.verbose. [ad 20081203]
	module(7): Add a layerfs module. [ad 20081205]
	libc: Add fdopendir(3), from Solaris/Linux. [ad 20081205]
	compat_linux32: Add getresuid() and getresgid() syscalls.
		[njoly 20081205]
	ddb(4): Add 'show vmem' and 'show all vmems' commands.
		[cegger 20081207]
	re(4): Add support for a 8168C/8111C revision. [alc 20081208]
	tab(1): Add tabs utility, fixes PR bin/11227. [roy 20081211]
	httpd(8): Add rc.d script to control httpd daemon. [mishka 20081211]
	ath(4): Switch from the binary HAL to the open-source HAL released by
		Sam Leffler. [alc 20081211]
	compat_linux32: Add clock_nanosleep syscall. [njoly 20081212]
	boot.cfg(5): Add an option to clear the screen. [christos 20081213]
	pms(4): Add Elantech multi-touch trackpad support [jmcneill 20081213]
	sparc: Switch to use wscons. [macallan 20081213]
	snprintb(3), kprintf(9): Share snprintb(3) among kernel and libutil
		and make it supersede bitmask_snprintf(9). [christos 20081216]
	sup(1): Add -M user flag to specify the username to send mail to from
		the command line. [christos 20081217]
	malloc(9): Remove MALLOC() and FREE() macros. [cegger 20081217]
	x68k:	pow(4) switches to use MI powerd(8) [isaki 20081220]
	viaide(4): Add VT8237S Integrated SATA Controller support.
		[nonaka 20081221]
	lvm(8):	Add LVM functionality to NetBSD; this is done by dm(4) driver
		in kernel and lvm2 tools made by RedHat. We use the same well
		known interface as Linux does for managing LVM. [haad 20081222]
	gcscaudio(4): Add AMD Geode CS5536 audio driver, from SHIMIZU Ryo.
		[jmcneill 20081228]
	rump: i386 and amd64 can use kernel module binaries [pooka 20081230]
	mount_cd9660(8): Add nocasetrans option as a synonym for nomaplcase.
		[christos 20090103]
	rump: Include libkern in rumpkern. [pooka 20090104]
	rc.d(8): Add a script for rndctl(8). [apb 20090104]
	module(7): Add sysvbfs module. [pooka 20090105]
	rump: Prepend all symbols within the rump virtual kernel image
		with the string "rump".  This prevents conflicts with the
		host system symbols.  [pooka 20090108]
	bwi(4): Add a driver for Broadcom AirForce / Apple Airport Extreme
		wireless network controllers. From Taylor Campbell.
		[macallan 20090109]
	rgephy(4): Support for RTL8211C(L) phy. From FreeBSD. [cegger 20090109]
	rgephy(4): Support for flow control. From OpenBSD. [cegger 20090109]
	makefs(8): Add support for encoding RISC OS metadata into
		ISO 9660 file systems. [bjh21 20090110]
	time_t, dev_t: Change time_t and dev_t to be 64 bit quantities. Fixed
		POSIX issues in timeval and timespec. [christos 20090110]
	make(1): Changes to the way .for loops are processed. [dsl 20090111]
	mount(8): Add -o rump, which mounts the file system as a userspace
		server via puffs and rump instead of using a
		kernel driver. [pooka 20090111]
	openpgpsdk: Initial import of openpgpsdk v0.9. [agc 20090112]
	rump: Implement a virtual block device driver.  [pooka 20090113]
	rump: Support __RENAME in syscall interfaces and expose the

lib/libisns/isns.c
/*	$NetBSD: isns.c,v 1.1.1.1 2011/01/16 01:22:50 agc Exp $	*/

/*-
 * Copyright (c) 2004,2009 The NetBSD Foundation, Inc.
 * All rights reserved.
 *
 * This code is derived from software contributed to The NetBSD Foundation
 * by Wasabi Systems, Inc.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include <sys/cdefs.h>
__RCSID("$NetBSD: isns.c,v 1.1.1.1 2011/01/16 01:22:50 agc Exp $");

/*
 * isns.c
 */

#include <sys/types.h>
#include <sys/param.h>
#include <sys/uio.h>
#include <sys/poll.h>

#include <errno.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>

#include "isns.h"
#include "isns_config.h"



/*
 * isns_init()
 */
int
isns_init(ISNS_HANDLE *isns_handle_p, int is_server)
{
	struct isns_config_s *cfg_p;
	int rval;

	*isns_handle_p = NULL;

	if ((cfg_p = isns_new_config()) == NULL) {
		DBG("isns_init: error on isns_new_config()\n");
		return ENOMEM;
	}

	cfg_p->is_server = is_server;
	cfg_p->curtask_p = NULL;

	if ((rval = pipe(cfg_p->pipe_fds)) != 0) {
		DBG("isns_init: error on wepe_sys_pipe()\n");
		isns_destroy_config(cfg_p);
		return rval;
	}

	if ((cfg_p->kq = kqueue()) == -1) {
		DBG("isns_init: error on kqueue()\n");
		isns_destroy_config(cfg_p);
		return -1;
	}

	rval = isns_change_kevent_list(cfg_p, (uintptr_t)cfg_p->pipe_fds[0],
	    EVFILT_READ, EV_ADD | EV_ENABLE, (int64_t)0,
	    (intptr_t)isns_kevent_pipe);
	if (rval == -1) {
		DBG("isns_init: error on isns_change_kevent_list() "
		    "for isns_kevent_pipe\n");
		isns_destroy_config(cfg_p);
		return rval;
	}

	isns_init_buffer_pool();
	rval = isns_add_buffer_pool(ISNS_BUF_SIZE, ISNS_BUF_COUNT);
	if (rval != 0) {
		DBG("isns_init: error on isns_init_buffer_pool()\n");
		isns_destroy_config(cfg_p);
		return rval;
	}
	rval = isns_add_buffer_pool((int)ISNS_SMALL_BUF_SIZE, ISNS_SMALL_BUF_COUNT);
	if (rval != 0) {
		DBG("isns_init: error on isns_init_buffer_pool() [small]\n");
		isns_destroy_config(cfg_p);
		isns_destroy_buffer_pool();
		return rval;
	}

	if ((rval = isns_thread_create(cfg_p)) != 0) {
		DBG("isns_init: error on isns_thread_create()\n");
		isns_destroy_config(cfg_p);
		isns_destroy_buffer_pool();
		return rval;
	}

	*isns_handle_p = (ISNS_HANDLE)cfg_p;

	return 0;
}


/*
 * isns_add_servercon()
 */
int
isns_add_servercon(ISNS_HANDLE isns_handle, int fd, struct addrinfo *ai)
{
	struct isns_config_s *cfg_p;
	struct isns_task_s *task_p;
	struct addrinfo *ai_p;
	size_t len;

	if (isns_handle == ISNS_INVALID_HANDLE)
		return EINVAL;

	cfg_p = (struct isns_config_s *)isns_handle;

	ai_p = (struct addrinfo *)isns_malloc(sizeof(struct addrinfo));
	if (ai_p == NULL)
		return ENOMEM;

	ai_p->ai_flags = ai->ai_flags;
	ai_p->ai_family = ai->ai_family;
	ai_p->ai_socktype = ai->ai_socktype;
	ai_p->ai_protocol = ai->ai_protocol;
	ai_p->ai_addrlen = ai->ai_addrlen;
	if (ai->ai_canonname != NULL) {
		len = strlen(ai->ai_canonname);
		ai_p->ai_canonname = (char *)isns_malloc(len + 1);
		if (ai_p->ai_canonname == NULL) {
			isns_free(ai_p);
			return ENOMEM;
		}
		memset(ai_p->ai_canonname, '\0', len + 1);
		strncpy(ai_p->ai_canonname, ai->ai_canonname, len);
	} else
		ai_p->ai_canonname = NULL;
	if (ai->ai_addr != NULL) { 
		ai_p->ai_addr = (struct sockaddr *)isns_malloc(ai_p->
		    ai_addrlen);
		if (ai_p->ai_addr == NULL) {
			if (ai_p->ai_canonname != NULL)
				isns_free(ai_p->ai_canonname);
			isns_free(ai_p);
			return ENOMEM;
		}
		memcpy(ai_p->ai_addr, ai->ai_addr, ai_p->ai_addrlen);
	} else
		ai_p->ai_addr = NULL;
	ai_p->ai_next = NULL;
	
	/* Build task and kick off task processing */
	task_p = isns_new_task(cfg_p, ISNS_TASK_INIT_SOCKET_IO, 1);
	task_p->var.init_socket_io.sd = fd;
	task_p->var.init_socket_io.ai_p = ai_p;

	isns_taskq_insert_head(cfg_p, task_p);
	isns_issue_cmd(cfg_p, ISNS_CMD_PROCESS_TASKQ);
	isns_wait_task(task_p, NULL);

	return 0;
}


/*
 * isns_init_reg_refresh()
 */
int
isns_init_reg_refresh(ISNS_HANDLE isns_handle, const char *node, int interval)
{
	struct isns_config_s *cfg_p;
	struct isns_task_s *task_p;
	struct isns_refresh_s *ref_p;

	if (isns_handle == ISNS_INVALID_HANDLE)
		return EINVAL;

	/* Build INIT_REFRESH task with info provided. */
	cfg_p = (struct isns_config_s *)isns_handle;
	task_p = isns_new_task(cfg_p, ISNS_TASK_INIT_REFRESH, 0);
	if (task_p == NULL)
		return ENOMEM;

	ref_p = (struct isns_refresh_s *)
	    isns_malloc(sizeof(struct isns_refresh_s));
	if (ref_p == NULL) {
		isns_free_task(task_p);
		return ENOMEM;
	}

	(void) snprintf(ref_p->node, sizeof(ref_p->node), "%.*s",
		(int)sizeof(ref_p->node)-1, node);
	ref_p->interval = interval;
	ref_p->trans_p = NULL;
	task_p->var.init_refresh.ref_p = ref_p;

	isns_taskq_insert_tail(cfg_p, task_p);
	isns_issue_cmd(cfg_p, ISNS_CMD_PROCESS_TASKQ);

	return 0;
}


/*
 * isns_stop()
 */
void isns_stop(ISNS_HANDLE isns_handle)
{
	struct isns_config_s *cfg_p = (struct isns_config_s *)isns_handle;

	DBG("isns_stop: entered\n");

	isns_issue_cmd(cfg_p, ISNS_CMD_STOP);

	isns_thread_destroy(cfg_p);
	isns_destroy_config(cfg_p);
	isns_destroy_buffer_pool();
}
